src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp

```diff
@@ -197,6 +197,80 @@ typename DiracDeterminantBatched<DET_ENGINE>::Grad DiracDeterminantBatched<DET_E
   return g;
 }
 
+template<typename DET_ENGINE>
+void DiracDeterminantBatched<DET_ENGINE>::mw_evalGradWithSpin(
+    const RefVectorWithLeader<WaveFunctionComponent>& wfc_list,
+    const RefVectorWithLeader<ParticleSet>& p_list,
+    int iat,
+    std::vector<Grad>& grad_now,
+    std::vector<ComplexType>& spingrad_now) const
+{
+  assert(this == &wfc_list.getLeader());
+  auto& wfc_leader = wfc_list.getCastedLeader<DiracDeterminantBatched<DET_ENGINE>>();
+  ScopedTimer local_timer(RatioTimer);
+
+  const int nw = wfc_list.size();
+  RefVectorWithLeader<SPOSet> phi_list(*Phi);
+  std::vector<SPOSet::ValueVector> psi_values, lap_values, dspin_values;
+  std::vector<SPOSet::GradVector> grad_values;
+  RefVector<SPOSet::ValueVector> psi_v_list, lap_v_list, dspin_v_list;
+  RefVector<SPOSet::GradVector> grad_v_list;
+
+  phi_list.reserve(nw);
+  psi_values.reserve(nw);
+  grad_values.reserve(nw);
+  lap_values.reserve(nw);
+  dspin_values.reserve(nw);
+
+  psi_v_list.reserve(nw);
+  grad_v_list.reserve(nw);
+  lap_v_list.reserve(nw);
+  dspin_v_list.reserve(nw);
+
+  SPOSet::ValueVector val(wfc_leader.Phi->size());
+  SPOSet::GradVector grad(wfc_leader.Phi->size());
+  for (int iw = 0; iw < wfc_list.size(); iw++)
+  {
+    auto& det = wfc_list.getCastedElement<DiracDeterminantBatched<DET_ENGINE>>(iw);
+    phi_list.push_back(*det.Phi);
+
+    psi_values.emplace_back(val);
+    grad_values.emplace_back(grad);
+    lap_values.emplace_back(val);
+    dspin_values.emplace_back(val);
+
+    psi_v_list.emplace_back(psi_values.back());
+    grad_v_list.emplace_back(grad_values.back());
+    lap_v_list.emplace_back(lap_values.back());
+    dspin_v_list.emplace_back(dspin_values.back());
+  }
+  auto& phi_leader = phi_list.getLeader();
+  phi_leader.mw_evaluateVGLWithSpin(phi_list, p_list, iat, psi_v_list, grad_v_list, lap_v_list, dspin_v_list);
```

Comment:
So  I guess this is all being done to get dspin_values via dspin_v_list?  I don't see where grad_values, lap_values etc. are ever used. There is no indication of in and out variables int the SPOSet docs.

Reply:
Yes, mw_evaluateVGLWithSpin was taking RefVectors as arguments, so I constructed the vectors of values first and then use emplace_back(xxx_values.back()) to add the references to the RefVector. 

All this is to recompute the spin gradients...and this approach is just reusing existing functionality to compute the spin gradients since we aren't storing them and updating them along the way. I was reusing the existing mw_evaluateVGLWithSpin since it already accomplished what I needed. If I didn't want to recalculate the GL and just have an API to give the spin gradients only, then inside SpinorSet I would need a mw_evaluateValue for each of the up/down SPOSets that make up the spinor in order to construct the spin gradient.  That API is defined in SPOSet, but it is just a simple loop over the legacy single walker APIs evaluateValue. There isn't a mw_evaluateValue in SplineC2C that utilizes the offload...so it just seemed simpler to use the existing API. 

I spoke with @ye-luo on slack about this...basically there are two approached to this. First is we store the spin gradients and then need to either add to the existing structures (like OffloadMWVGLArray for example) or add new ones to accommodate for the extra spin gradients...and then need to update the values whenever we call mw_accept_rejectMove. The other option (the one chosen here) is to just recompute the spingradients from the SPOSet from scratch every time we call TrialWaveFunction::mw_evalGrad and the spin gradients are being used. Ye thought there may not be a significant penalty to just recomputing, and it seemed simpler so I tried that here. 

I would actually prefer the first approach where we store and update when we accept/reject moves, but I was having trouble deciding the best way to store the spin gradients so that we only allocate on the device when we need it =

Reply:
Doing this recompute is quite a bit simpler than the other option. I've been working on a version where I store the spin gradients and update with accept move, and that has required a lot more code changes the way I've been doing it. 

It looks like the recalculation of things that aren't being used was also done elsewhere in this file...see DiracDeterminantBatched::mw_calcRatio. There is a call to Phi->mw_evaluateVGLandDetRatioGrads even though an implementation of something like mw_evaluateVandDetRatio would be more approriate so there isn't any computation of things not being used.